pkg/pipelinerunmetrics/metrics.go

```diff
@@ -272,9 +294,21 @@ func (r *Recorder) RunningPipelineRuns(lister listers.PipelineRunLister) error {
 	}
 
 	var runningPRs int
+	var trsWaitResolvingTaskRef int
+	var prsWaitResolvingPipelineRef int
+
 	for _, pr := range prs {
```

Comment:
I can see that we're already doing things this way for the number of running PRs, but I'm confused how this works. Is `RunningPipelineRuns` called every 30s, as indicated by the comment below? Would this increment a global constant every time `RunningPipelineRuns` is called? Do we ever need to decrement the counter when resolution is complete?

Reply:
Our experience with the throttled by quota / node resources I added back in May, i.e. following this prior "running implementation pattern" as you noted, is that the `metrics.Record` calls below act as a "set" operation (and hence resets every 30s) vs. an "increment" operation.

So if say `trsWaitResolvingTaskRef` or `prsWaitResolvingPipelineRef` is 0, the `metrics.Record` call below sets the metric to 0 and we effectively "decrements" or "resets".

For example, I'm looking at a query of `running_taskruns_throttled_by_node_count` in our cluster today, and it went up as high as 18 at 5:57 AM US Eastern, and by 6:30 had gone back down to zero as resources became available and those items got processed.

Now, as I read the code, and what I see in our running deployment, part of what makes this work is that theses "running" metrics are essentially "global" gauges, as none of the optional labels are applied (like is possible with the duration and latency metrics).

Hence the "reset" works regardless if say pipelineruns/taskruns have been deleted or are no longer running.

Now the "non-running" metrics with the optional labels are not managed in this "running" path, but rather inline in the various reconciliation paths.  And you'll see decrements there.

Hopefully I'm getting all the nuances of your point @lbernick and what I conveyed lines up with you.  Of course let me know if more discussion is needed.

Thanks

Reply:
got it, thanks for the explanation!